homeassistant/components/media_player/onkyo.py

```diff
@@ -9,37 +9,67 @@
 from homeassistant.components.media_player import (
     SUPPORT_TURN_OFF, SUPPORT_TURN_ON, SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET,
     SUPPORT_SELECT_SOURCE, MediaPlayerDevice)
-from homeassistant.const import STATE_OFF, STATE_ON
+from homeassistant.const import STATE_OFF, STATE_ON, CONF_HOST, CONF_NAME
 
 REQUIREMENTS = ['https://github.com/danieljkemp/onkyo-eiscp/archive/'
                 'python3.zip#onkyo-eiscp==0.9.2']
 _LOGGER = logging.getLogger(__name__)
 
 SUPPORT_ONKYO = SUPPORT_VOLUME_SET | SUPPORT_VOLUME_MUTE | \
     SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE
+KNOWN_HOSTS = []
+DEFAULT_SOURCES = {"TV": "tv", "Bluray": "bd", "Game": "game", "Aux1": "aux1",
+                   "Video 1": "video1", "Video 2": "video2",
+                   "Video 3": "video3", "Video 4": "video4",
+                   "Video 5": "video5", "Video 6": "video6",
+                   "Video 7": "video7"}
+CONFIG_SOURCE_LIST = "sources"
 
 
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Setup the Onkyo platform."""
+    import eiscp
     from eiscp import eISCP
-    add_devices(OnkyoDevice(receiver)
-                for receiver in eISCP.discover())
-
-
+    hosts = []
+
+    if CONF_HOST in config and config[CONF_HOST] not in KNOWN_HOSTS:
+        try:
+            hosts.append(OnkyoDevice(eiscp.eISCP(config[CONF_HOST]),
+                                     config[CONF_NAME],
+                                     config.get(CONFIG_SOURCE_LIST,
+                                                DEFAULT_SOURCES)))
+            KNOWN_HOSTS.append(config[CONF_HOST])
+        except OSError:
+            _LOGGER.error('Unable to connect to receiver at %s.',
+                          config[CONF_HOST])
+    else:
+        for receiver in eISCP.discover():
+            if receiver.host not in KNOWN_HOSTS:
+                hosts.append(OnkyoDevice(receiver,
+                                         sources=config[CONFIG_SOURCE_LIST]))
+                KNOWN_HOSTS.append(receiver.host)
+    add_devices(hosts)
+
+
+# pylint: disable=too-many-instance-attributes
 class OnkyoDevice(MediaPlayerDevice):
     """Representation of a Onkyo device."""
 
     # pylint: disable=too-many-public-methods, abstract-method
-    def __init__(self, receiver):
+    def __init__(self, receiver, name=None, sources=None):
         """Initialize the Onkyo Receiver."""
         self._receiver = receiver
         self._muted = False
         self._volume = 0
         self._pwstate = STATE_OFF
-        self.update()
-        self._name = '{}_{}'.format(
+        self._name = name or '{}_{}'.format(
             receiver.info['model_name'], receiver.info['identifier'])
         self._current_source = None
+        self._source_list = list(sources.keys() or DEFAULT_SOURCES.keys())
```

Comment:
I think that you should swap `sources` for `DEFAULT_SOURCES` in the `setup_platform`. Now the next two lines will use a different dictionary if `sources` is empty! If `sources` is `None` you'll get an attributeError